### PR TITLE
Implemented string builder pooling

### PR DESCRIPTION
**Pull request by mkrueger**
# Description

Object pooling for string builders and collection can reduce memory allocations/deallocations. This PR is for discussing the pros & cons of the approach. Is it useful to do?

# All SDK Contribution checklist:
- [x ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/pooling/PooledStringBuilder.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+    public static PooledStringBuilder createInstance() {
+        return s_poolInstance.get();
+    }
+    public static ObjectPool<PooledStringBuilder> createPool() {
+        return createPool(32);
+    }
+
+    public static ObjectPool<PooledStringBuilder> createPool(int size) {
+        return new ObjectPool<PooledStringBuilder>((pool) -> new PooledStringBuilder(pool), size);
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/pooling/PooledStringBuilder.java:23]: We are creating a new `ObjectFactory` every single time. Why can't we use the same factory instance?
What is the limit on number of `ObjectFactory` instances that can be created?

**Review by kushagraThapar — CHANGES_REQUESTED** [review]

**Comment by kushagraThapar** [comment]
@mkrueger - the overall approach is definitely worth discussing, however it will be good to get some data on performance / memory benchmarking to see what are the effects of this before we dive too deep into technical discussion on the PR.

**Comment by mkrueger** [comment]
I did some micro benchmarking it was quite good ~2x. Did it today and only showed a small advantage and even a worse in another use case.

Can't really explain that yet - but even in the case where it's faster the advantage is not enough to justify the change. I attached the demo project:

[JavaCollections.zip](https://github.com/Azure/azure-sdk-for-java/files/12038531/JavaCollections.zip)

**Comment by mkrueger** [comment]
Closing for now.

**Pull request closed without merging**